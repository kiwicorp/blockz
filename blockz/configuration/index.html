<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Support for common configuration behaviour and loading them from various sources."><meta name="keywords" content="rust, rustlang, rust-lang, configuration"><title>blockz::configuration - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../blockz/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Module configuration</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li></ul></div><div id="sidebar-vars" data-name="configuration" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img src="../../brush.svg" width="18" height="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img src="../../wheel.svg" width="18" height="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">blockz</a>::<wbr><a class="mod" href="#">configuration</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/blockz/configuration.rs.html#1-274" title="goto source code">[src]</a></span></h1><div class="item-info"><div class="stab portability">This is supported on <strong>crate feature <code>configuration</code></strong> only.</div></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Support for common configuration behaviour and loading them from various sources.</p>
<h1 id="direct-configuration-example" class="section-header"><a href="#direct-configuration-example">Direct configuration example</a></h1>
<p>This example shows how to use the direct configuration. Please note that the direct
configuration is usually useful in testing scenarios. Production usage is discouraged.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Configuration</span>)]</span>
<span class="attribute">#[<span class="ident">configuration</span>(<span class="ident">direct</span>)]</span>
<span class="kw">struct</span> <span class="ident">ServerConfig</span> {
    <span class="ident">address</span>: <span class="ident">String</span>,
    <span class="ident">port</span>: <span class="ident">u16</span>,
}

<span class="kw">let</span> <span class="ident">config</span> <span class="op">=</span> <span class="op">&lt;</span><span class="ident">ServerConfig</span> <span class="kw">as</span> <span class="ident">Configuration</span><span class="op">&gt;</span><span class="ident">::load</span>(<span class="ident">ServerConfig</span> {
    <span class="ident">address</span>: <span class="string">&quot;127.0.0.1&quot;</span>.<span class="ident">to_string</span>(),
    <span class="ident">port</span>: <span class="number">58732</span>,
}).<span class="kw">await</span>;
<span class="macro">println!</span>(<span class="string">&quot;Server binding to {}:{}.&quot;</span>, <span class="ident">config</span>.<span class="ident">address</span>, <span class="ident">config</span>.<span class="ident">port</span>);
<span class="comment">// Server binding to 127.0.0.1:58732.</span></pre></div>
<p>If the configuration type implements Default, you can use EasyConfiguration to load the default
value, like so:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Configuration</span>)]</span>
<span class="attribute">#[<span class="ident">configuration</span>(<span class="ident">direct</span>)]</span>
<span class="kw">struct</span> <span class="ident">ServerConfig</span> {
    <span class="ident">address</span>: <span class="ident">String</span>,
    <span class="ident">port</span>: <span class="ident">u16</span>,
}

<span class="kw">impl</span> <span class="ident">Default</span> <span class="kw">for</span> <span class="ident">ServerConfig</span> {
    <span class="kw">fn</span> <span class="ident">default</span>() <span class="op">-</span><span class="op">&gt;</span> <span class="ident">ServerConfig</span> {
        <span class="ident">ServerConfig</span> {
            <span class="ident">address</span>: <span class="string">&quot;0.0.0.0&quot;</span>.<span class="ident">to_string</span>(),
            <span class="ident">port</span>: <span class="number">9999</span>,
        }
    }
}

<span class="kw">let</span> <span class="ident">config</span> <span class="op">=</span> <span class="op">&lt;</span><span class="ident">ServerConfig</span> <span class="kw">as</span> <span class="ident">EasyConfiguration</span><span class="op">&gt;</span><span class="ident">::load</span>().<span class="kw">await</span>;
<span class="macro">println!</span>(<span class="string">&quot;Server binding to {}:{}.&quot;</span>, <span class="ident">config</span>.<span class="ident">address</span>, <span class="ident">config</span>.<span class="ident">port</span>);
<span class="comment">// Server binding to 0.0.0.0:9999.</span></pre></div>
<h1 id="env-configuration-examples" class="section-header"><a href="#env-configuration-examples">Env configuration examples</a></h1>
<p><strong>NOTE</strong>: These examples require the <strong>env_configuration</strong> feature.</p>
<p>The first example shows how to use the env configuration with a prefix.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Configuration</span>, <span class="ident">Deserialize</span>)]</span>
<span class="attribute">#[<span class="ident">configuration</span>(<span class="ident">env</span>(<span class="ident">prefix</span> <span class="op">=</span> <span class="string">&quot;COOL_APP_&quot;</span>))]</span>
<span class="kw">struct</span> <span class="ident">ServerConfig</span> {
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_addr&quot;</span>)]</span>
    <span class="ident">address</span>: <span class="ident">String</span>,
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_port&quot;</span>)]</span>
    <span class="ident">port</span>: <span class="ident">u16</span>,
}

<span class="kw">let</span> <span class="ident">config</span> <span class="op">=</span> <span class="op">&lt;</span><span class="ident">ServerConfig</span> <span class="kw">as</span> <span class="ident">EasyConfiguration</span><span class="op">&gt;</span><span class="ident">::load</span>()
    .<span class="kw">await</span>
    .<span class="ident">expect</span>(<span class="string">&quot;Failed to load configuration from the environment!&quot;</span>);
<span class="macro">println!</span>(<span class="string">&quot;Server binding to {}:{}.&quot;</span>, <span class="ident">config</span>.<span class="ident">address</span>, <span class="ident">config</span>.<span class="ident">port</span>);
<span class="comment">// Server binding to aaa.bbb.ccc.ddd:ppppp.</span></pre></div>
<p>The second example shows how to use the env configuration with a prefix source in the form of a
constant.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="doccomment">/// This is our prefix.</span>
<span class="kw">const</span> <span class="ident">PREFIX</span>: <span class="kw-2">&amp;</span><span class="ident">str</span> <span class="op">=</span> <span class="string">&quot;COOL_APP_&quot;</span>;

<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Configuration</span>, <span class="ident">Deserialize</span>)]</span>
<span class="attribute">#[<span class="ident">configuration</span>(<span class="ident">env</span>(<span class="ident">prefix_source</span> <span class="op">=</span> <span class="string">&quot;PREFIX.to_string()&quot;</span>))]</span>
<span class="kw">struct</span> <span class="ident">ServerConfig</span> {
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_addr&quot;</span>)]</span>
    <span class="ident">address</span>: <span class="ident">String</span>,
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_port&quot;</span>)]</span>
    <span class="ident">port</span>: <span class="ident">u16</span>,
}

<span class="kw">let</span> <span class="ident">config</span> <span class="op">=</span> <span class="op">&lt;</span><span class="ident">ServerConfig</span> <span class="kw">as</span> <span class="ident">EasyConfiguration</span><span class="op">&gt;</span><span class="ident">::load</span>()
    .<span class="kw">await</span>
    .<span class="ident">expect</span>(<span class="string">&quot;Failed to load configuration from the environment!&quot;</span>);
<span class="macro">println!</span>(<span class="string">&quot;Server binding to {}:{}.&quot;</span>, <span class="ident">config</span>.<span class="ident">address</span>, <span class="ident">config</span>.<span class="ident">port</span>);
<span class="comment">// Server binding to aaa.bbb.ccc.ddd:ppppp.</span></pre></div>
<p>The third example shows how to use the env configuration with a prefix source in the form of a
function that can return an error. The error must be mapped to an <code>envy::Error</code>.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="doccomment">/// This function sources our prefix.</span>
<span class="kw">fn</span> <span class="ident">source_prefix</span>() <span class="op">-</span><span class="op">&gt;</span> <span class="prelude-ty">Result</span><span class="op">&lt;</span><span class="ident">String</span>, <span class="ident">envy::Error</span><span class="op">&gt;</span> {
    <span class="ident">std::env::var</span>(<span class="string">&quot;COOL_APP_PREFIX&quot;</span>)
        .<span class="ident">map_err</span>(<span class="op">|</span><span class="ident">err</span><span class="op">|</span> <span class="ident">envy::Error::Custom</span>(<span class="macro">format!</span>(<span class="string">&quot;failed to source the prefix: {}&quot;</span>, <span class="ident">err</span>)))
}

<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Configuration</span>, <span class="ident">Deserialize</span>)]</span>
<span class="attribute">#[<span class="ident">configuration</span>(<span class="ident">env</span>(<span class="ident">prefix_source</span> <span class="op">=</span> <span class="string">&quot;source_prefix()?&quot;</span>))]</span>
<span class="kw">struct</span> <span class="ident">ServerConfig</span> {
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_addr&quot;</span>)]</span>
    <span class="ident">address</span>: <span class="ident">String</span>,
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_port&quot;</span>)]</span>
    <span class="ident">port</span>: <span class="ident">u16</span>,
}

<span class="kw">let</span> <span class="ident">config</span> <span class="op">=</span> <span class="op">&lt;</span><span class="ident">ServerConfig</span> <span class="kw">as</span> <span class="ident">EasyConfiguration</span><span class="op">&gt;</span><span class="ident">::load</span>()
    .<span class="kw">await</span>
    .<span class="ident">expect</span>(<span class="string">&quot;Failed to load configuration from the environment!&quot;</span>);
<span class="macro">println!</span>(<span class="string">&quot;Server binding to {}:{}.&quot;</span>, <span class="ident">config</span>.<span class="ident">address</span>, <span class="ident">config</span>.<span class="ident">port</span>);
<span class="comment">// Server binding to aaa.bbb.ccc.ddd:ppppp.</span></pre></div>
</div></details><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<table><tr class="module-item"><td><a class="struct" href="struct.DirectConfiguration.html" title="blockz::configuration::DirectConfiguration struct">DirectConfiguration</a></td><td class="docblock-short"><p>Direct configuration that just returns the passed value.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.EnvConfiguration.html" title="blockz::configuration::EnvConfiguration struct">EnvConfiguration</a></td><td class="docblock-short"><span class="stab portability" title="This is supported on crate feature `env_configuration` only"><code>env_configuration</code></span><p>Configuration that can be sourced from environment variables.</p>
</td></tr></table><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<table><tr class="module-item"><td><a class="trait" href="trait.Configuration.html" title="blockz::configuration::Configuration trait">Configuration</a></td><td class="docblock-short"><p>Common behaviour of configurations.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.EasyConfiguration.html" title="blockz::configuration::EasyConfiguration trait">EasyConfiguration</a></td><td class="docblock-short"><p>An easy configuration is a configuration that can be loaded without other parameters.</p>
</td></tr></table></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="blockz" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div><script src="../../main.js"></script></body></html>