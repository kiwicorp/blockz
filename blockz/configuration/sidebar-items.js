initSidebarItems({"struct":[["DirectConfiguration","Direct configuration that just returns the passed value."],["EnvConfiguration","Configuration that can be sourced from environment variables."]],"trait":[["Configuration","Common behaviour of configurations."],["EasyConfiguration","An easy configuration is a configuration that can be loaded without other parameters."]]});