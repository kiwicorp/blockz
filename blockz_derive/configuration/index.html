<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="#[derive(Configuration)]."><meta name="keywords" content="rust, rustlang, rust-lang, configuration"><title>blockz_derive::configuration - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../blockz_derive/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Module configuration</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#types">Type Definitions</a></li></ul></div><div id="sidebar-vars" data-name="configuration" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img src="../../brush.svg" width="18" height="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img src="../../wheel.svg" width="18" height="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">blockz_derive</a>::<wbr><a class="mod" href="#">configuration</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/blockz_derive/configuration/mod.rs.html#1-120" title="goto source code">[src]</a></span></h1><div class="item-info"><div class="stab portability">This is supported on <strong>crate feature <code>configuration</code></strong> only.</div></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>#[derive(Configuration)].</p>
</div></details><h2 id="modules" class="section-header"><a href="#modules">Modules</a></h2>
<table><tr class="module-item"><td><a class="mod" href="direct/index.html" title="blockz_derive::configuration::direct mod">direct</a></td><td class="docblock-short"><p>Direct configuration factory.</p>
</td></tr><tr class="module-item"><td><a class="mod" href="env/index.html" title="blockz_derive::configuration::env mod">env</a></td><td class="docblock-short"><span class="stab portability" title="This is supported on crate feature `env_configuration` only"><code>env_configuration</code></span><p>Env configuration factory.</p>
</td></tr></table><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<table><tr class="module-item"><td><a class="struct" href="struct.ConfigurationFactory.html" title="blockz_derive::configuration::ConfigurationFactory struct">ConfigurationFactory</a></td><td class="docblock-short"><p>A factory that builds implementations for the Configuration trait.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.ConfigurationOpts.html" title="blockz_derive::configuration::ConfigurationOpts struct">ConfigurationOpts</a></td><td class="docblock-short"><p>Options used by the configuration factory.</p>
</td></tr></table><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<table><tr class="module-item"><td><a class="type" href="type.DynFactory.html" title="blockz_derive::configuration::DynFactory type">DynFactory</a></td><td class="docblock-short"><p>A dynamic factory that produces a token stream.</p>
</td></tr><tr class="module-item"><td><a class="type" href="type.DynFactoryProduct.html" title="blockz_derive::configuration::DynFactoryProduct type">DynFactoryProduct</a></td><td class="docblock-short"><p>A product produced by a dyn factory.</p>
</td></tr><tr class="module-item"><td><a class="type" href="type.FnNewDynFactory.html" title="blockz_derive::configuration::FnNewDynFactory type">FnNewDynFactory</a></td><td class="docblock-short"><p>A function that creates a specialized factory.</p>
</td></tr></table></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="blockz_derive" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div><script src="../../main.js"></script></body></html>