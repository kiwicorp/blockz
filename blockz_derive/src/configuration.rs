//! #[derive(Configuration)].

use proc_macro2::Ident;
use proc_macro2::TokenStream;

#[cfg(feature = "no_absolute_paths")]
pub fn impl_configuration_trait(type_name: &Ident, prefix: Option<LitStr>) -> TokenStream {
    quote! {
        #[async_trait::async_trait]
        impl ::blockz::configuration::Configuration for #type_name {
            type Inner = #type_name;
            type Opts = Option<String>;

            async fn load(opts: Self::Opts) -> ::anyhow::Result<Self::Inner> {
                if let Some(prefix) = opts {
                    Ok(::envy::prefixed(prefix).from_env::<Self::Inner>()?)
                } else {
                    Ok(::envy::from_env::<Self::Inner>()?)
                }
            }
        }
    }
}

#[cfg(not(feature = "no_absolute_paths"))]
pub fn impl_configuration_trait(type_name: &Ident) -> TokenStream {
    quote! {
        #[async_trait::async_trait]
        impl blockz::configuration::Configuration for #type_name {
            type Inner = #type_name;
            type Opts = Option<String>;

            async fn load(opts: Self::Opts) -> anyhow::Result<Self::Inner> {
                if let Some(prefix) = opts {
                    Ok(envy::prefixed(prefix).from_env::<Self::Inner>()?)
                } else {
                    Ok(envy::from_env::<Self::Inner>()?)
                }
            }
        }
    }
}
