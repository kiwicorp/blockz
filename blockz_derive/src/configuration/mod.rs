//! #[derive(Configuration)].

mod direct;
#[cfg(feature = "env_configuration")]
#[cfg_attr(docsrs, doc(cfg(feature = "env_configuration")))]
mod env;

use darling::FromDeriveInput;

use proc_macro2::TokenStream;

use syn::DeriveInput;

use crate::factory::Factory;
use crate::factory::ReusableFactory;
use crate::ProcMacroErrorExt;

use self::direct::DirectConfigurationFactory;
#[cfg(feature = "env_configuration")]
use self::env::EnvConfigurationFactory;
#[cfg(feature = "env_configuration")]
use self::env::EnvConfigurationOpts;

/// A product produced by a dyn factory.
type DynFactoryProduct = Result<TokenStream, Box<dyn ProcMacroErrorExt>>;

/// A dynamic factory that produces a token stream.
type DynFactory = Box<dyn ReusableFactory<Product = DynFactoryProduct>>;

/// A function that creates a specialized factory.
type FnNewDynFactory = Box<dyn FnOnce(DeriveInput, &mut ConfigurationOpts) -> DynFactory>;

/// A macro that automatically builds the functions for selecting the dynamic factory used for
/// producing the macro output.
macro_rules! feature_factory {
    ($feature_name: literal, $function_name: ident, $opts_field: ident, $factory_new_dyn: path) => {
        #[allow(unused_variables)]
        fn $function_name(opts: Option<&ConfigurationOpts>) -> Option<FnNewDynFactory> {
            #[cfg(not(feature = $feature_name))]
            {
                None
            }
            #[cfg(feature = $feature_name)]
            {
                let opts = if let Some(value) = opts {
                    value
                } else {
                    return Some(Box::new($factory_new_dyn));
                };
                if opts.$opts_field.is_none() {
                    None
                } else {
                    Some(Box::new($factory_new_dyn))
                }
            }
        }
    };
}

/// A factory that builds implementations for the Configuration trait.
pub(crate) struct ConfigurationFactory {
    input: DeriveInput,
    opts: ConfigurationOpts,
}

/// Options used by the configuration factory.
#[derive(FromDeriveInput)]
#[darling(attributes(configuration))]
struct ConfigurationOpts {
    #[cfg(feature = "env_configuration")]
    #[cfg_attr(feature = "env_configuration", darling(default))]
    env: Option<EnvConfigurationOpts>,
    #[darling(default)]
    direct: bool,
}

impl ConfigurationFactory {
    /// Create a new configuration factory.
    pub fn new(input: DeriveInput) -> Result<Self, darling::Error> {
        ConfigurationOpts::from_derive_input(&input).map(|opts| Self { input, opts })
    }

    fn direct_confiuration_factory(opts: &ConfigurationOpts) -> Option<FnNewDynFactory> {
        if opts.direct {
            Some(Box::new(DirectConfigurationFactory::new_dyn))
        } else {
            None
        }
    }

    feature_factory!(
        "env_configuration",
        env_configuration_factory,
        env,
        EnvConfigurationFactory::new_dyn
    );

    fn pick_new_dyn_factory_fn(opts: Option<&ConfigurationOpts>) -> Option<FnNewDynFactory> {
        Self::env_configuration_factory(opts).or_else(Option::default)
    }

    /// Returns a function that creates the new dynamic factory or None if the options did not
    /// specify a preference.
    fn get_new_dyn_factory_fn(&self) -> FnNewDynFactory {
        Self::pick_new_dyn_factory_fn(Some(&self.opts))
            .or_else(|| Self::direct_confiuration_factory(&self.opts))
            .or_else(|| Self::pick_new_dyn_factory_fn(None))
            .unwrap_or(Box::new(DirectConfigurationFactory::new_dyn))
    }
}

impl Factory for ConfigurationFactory {
    type Product = DynFactoryProduct;

    fn build(mut self) -> Self::Product {
        let factory_builder: FnNewDynFactory = self.get_new_dyn_factory_fn();
        let mut factory = factory_builder(self.input, &mut self.opts);
        factory.build()
    }
}
