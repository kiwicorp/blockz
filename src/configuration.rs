//! Support for common configuration behaviour and loading them from various sources.
//!
//! # Example
//!
//! This example shows how to automatically derive Configuration on a struct. This particular
//! example will showcase loading the configuration from environment variables with a particular
//! prefix.
//!
//! **NOTE**: This example requires the __env_configuration__ feature.
//!
//! ```
//! use blockz::prelude::*;
//!
//! use serde::Deserialize;
//!
//! #[derive(Configuration, Deserialize)]
//! #[configuration(env(prefix = "COOL_APP_"))]
//! struct ServerConfig {
//!     #[serde(rename = "bind_addr")]
//!     address: String,
//!     #[serde(rename = "bind_port")]
//!     port: u16,
//! }
//!
//! #[tokio::main]
//! async fn main() {
//!     # std::env::set_var("COOL_APP_BIND_ADDR", "0.0.0.0");
//!     # std::env::set_var("COOL_APP_BIND_PORT", "58732");
//!     let config = <ServerConfig as EasyConfiguration>::load()
//!         .await
//!         .expect("Failed to load configuration from the environment!");
//!     # assert_eq!(config.address.as_str(), "0.0.0.0");
//!     # assert_eq!(config.port, 58732);
//!     println!("Server binding to {}:{}.", config.address, config.port);
//!     // Server binding to 0.0.0.0:58732.
//! }
//! ```


#[cfg(feature = "env_configuration")]
use serde::Deserialize;

use std::future::Future;
use std::marker::PhantomData;

/// Common behaviour of configurations.
#[async_trait::async_trait]
pub trait Configuration {
    /// The type of options container this Configuration accepts for the purpose of loading the
    /// configuration.
    type Opts: Send;

    /// The result type that can be produced by loading the configuration.
    type Result: Send;

    /// Load the configuration.
    async fn load(opts: Self::Opts) -> Self::Result;
}

/// An easy configuration is a configuration that can be loaded without other parameters.
#[async_trait::async_trait]
pub trait EasyConfiguration {
    /// The result type that can be produced by loading the configuration.
    type Result: Send;

    /// Load the configuration.
    async fn load() -> Self::Result;
}

/// Automatically implement EasyConfiguration on Configuration implementations whose Opts type is
/// Default.
#[async_trait::async_trait]
impl<C, O, R> EasyConfiguration for C
where
    C: Configuration<Opts = O, Result = R>,
    O: Default + Send + 'static,
    R: Send + 'static,
{
    type Result = R;

    async fn load() -> Self::Result {
        C::load(O::default()).await
    }
}

/// Direct configuration that just returns the passed value.
pub struct DirectConfiguration<T>
where
    T: Send,
{
    _phantom: PhantomData<T>,
}

#[async_trait::async_trait]
impl<T> Configuration for DirectConfiguration<T>
where
    T: Send,
{
    type Opts = T;
    type Result = T;

    async fn load(opts: Self::Opts) -> Self::Result {
        opts
    }
}

/// Configuration that can be sourced from environment variables.
#[cfg(feature = "env_configuration")]
#[cfg_attr(docsrs, doc(cfg(feature = "env_configuration")))]
pub struct EnvConfiguration<T>
where
    T: for<'de> Deserialize<'de> + Send,
{
    _phantom_t: PhantomData<T>,
}

#[cfg(feature = "env_configuration")]
#[cfg_attr(docsrs, doc(cfg(feature = "env_configuration")))]
#[async_trait::async_trait]
impl<T> Configuration for EnvConfiguration<T>
where
    T: for<'de> Deserialize<'de> + Send,
{
    type Opts = Option<String>;
    type Result = Result<T, envy::Error>;

    async fn load(opts: Self::Opts) -> Self::Result {
        if let Some(prefix) = opts {
            envy::prefixed(prefix).from_env::<T>()
        } else {
            envy::from_env::<T>()
        }
    }
}
