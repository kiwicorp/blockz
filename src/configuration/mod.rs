//! Configurations.

/// Common behaviour of configurations.
#[async_trait::async_trait]
pub trait Configuration {
    /// The inner type of the configuration that can be loaded.
    type Inner: Send;

    /// The type of options container this Configuration accepts for the purpose of loading the
    /// configuration.
    type Opts: Send;

    /// Load the configuration.
    async fn load(opts: Self::Opts) -> anyhow::Result<Self::Inner>;
}

#[cfg(feature = "env")]
mod env;

#[cfg(feature = "json")]
mod json;
