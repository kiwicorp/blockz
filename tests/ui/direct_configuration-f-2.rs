//! Direct configuration ui test #1 - no Default impl.

#![cfg(not(all(feature = "envy_configuration")))]

use blockz::configuration::EasyConfiguration;
use blockz::prelude::*;

#[derive(Configuration, PartialEq)]
struct EnvConfig {
    server_port: u32,
}

#[tokio::main]
async fn main() {
    let env_config = EnvConfig { server_port: 53812 };
    let conf1 = <EnvConfig as EasyConfiguration>::load().await.unwrap();
    panic!("Execution should not arrive here!");
}
