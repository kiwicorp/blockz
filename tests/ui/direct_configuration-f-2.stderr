error[E0277]: the trait bound `EnvConfig: Default` is not satisfied
  --> $DIR/direct_configuration-f-2.rs:16:17
   |
16 |     let conf1 = <EnvConfig as EasyConfiguration>::load().await.unwrap();
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `EnvConfig`
   |
   = note: required because of the requirements on the impl of `EasyConfiguration` for `EnvConfig`

error[E0277]: the trait bound `EnvConfig: Default` is not satisfied
  --> $DIR/direct_configuration-f-2.rs:16:17
   |
16 |     let conf1 = <EnvConfig as EasyConfiguration>::load().await.unwrap();
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `EnvConfig`
   |
  ::: $WORKSPACE/src/configuration.rs
   |
   |     async fn load() -> Result<Self::Inner, Self::Error>;
   |              ---- required by this bound in `blockz::configuration::EasyConfiguration::load`
   |
   = note: required because of the requirements on the impl of `EasyConfiguration` for `EnvConfig`
